src/profile.c

```diff
@@ -0,0 +1,211 @@
+        jl_ptls_t ptls2 = jl_all_tls_states[t_i];
+
+        // Take a look at the malloc'ed arrays for this thread
+        mallocarray_t *ma = ptls2->heap.mallocarrays;
+
+        // Zoom through seeing if the given pointer matches this array's data pointer
+        while (ma != NULL) {
+            if (ma->a->data == adata) {
+                // If it matches, re-enable the GC and return that value
+                jl_gc_enable(prev_value);
+                return ma->a;
+            }
+            ma = ma->next;
+        }
+    }
+
+    // We were unable to find it. :(
+    jl_gc_enable(prev_value);
+    return NULL;
+}
+
+JL_DLLEXPORT void jl_memprofile_track_alloc(void *v, uint16_t tag, size_t allocsz) JL_NOTSAFEPOINT
+{
+    // Filter out this call with our tag filter
+    if ((tag & memprof_tag_filter) != tag)
+        return;
+
+    // Store the current backtrace location into our buffer, and increment the
+    // buffer index by the number of elements added.
+    size_t bt_size_step = 0;
+    int incomplete = rec_backtrace((jl_bt_element_t*) bt_data_prof + bt_size_cur,
+                                   &bt_size_step, bt_size_max - bt_size_cur - 1,
+                                   0, 1);
+
+    // If we overran this buffer or don't have the place to store the allocation info frame,
+    // then don't record the memory trace and quit.
+    size_t alloc_size_step = 7;
+    if (incomplete || bt_size_cur + bt_size_step
+                                  + alloc_size_step >= bt_size_max) {
+        bt_overflow |= JL_MEMPROF_BT_OVERFLOW;
+        jl_memprofile_stop();
+        return;
+    }
+
+    // Store an allocation information frame into the buffer
+    uintptr_t entry_tags = jl_bt_entry_descriptor(1, 4, JL_BT_ALLOCINFO_FRAME_TAG, 0);
+    jl_bt_element_t *bt_entry = (jl_bt_element_t*) bt_data_prof + bt_size_cur + bt_size_step;
+    bt_entry[0].uintptr = JL_BT_NON_PTR_ENTRY;
+    bt_entry[1].uintptr = entry_tags;
+    // The location of the type information for this chunk of memory (tracked value).
+    // Initially set to nothing, populated later by `jl_memprofile_set_typeof()`
+    bt_entry[2].jlvalue = (jl_value_t*)jl_nothing;
+    // The location of the data in memory, used to match allocations with deallocations.
+    bt_entry[3].uintptr = (uintptr_t) v;
+    // The time at which this happened
+    bt_entry[4].uintptr = jl_clock_now();   // FIXME: double to a potentially 32-bit uintptr
+    // The size of the allocation, or 0 if this was a free.
+    bt_entry[5].uintptr = allocsz;
+    // Used to "tag" this allocation within a particular domain (CPU, GPU, other)
+    // or within a particular allocator (Pool, std, malloc), or as a free instead.
+    bt_entry[6].uintptr = tag;
+
+    // Add a NULL-separator
+    bt_size_cur += bt_size_step + alloc_size_step;
+    bt_data_prof[bt_size_cur++].uintptr = 0;
+    assert(bt_size_cur < bt_size_max);
+}
+
+JL_DLLEXPORT void jl_memprofile_set_typeof(void * v, void * ty) JL_NOTSAFEPOINT
```

Comment:
The fact that `jl_memprofile_set_typeof` is done separately from `jl_memprofile_track_alloc` makes this non-atomic with respect to any locking scheme we want to use within these functions. TBH that seems likely to lead to problems and it would be great to figure out how to make writing the allocation trace and info more atomic.